engine/src/main/java/org/terasology/world/chunks/remoteChunkProvider/RemoteChunkProvider.java

```diff
@@ -24,105 +11,100 @@
+import org.terasology.world.chunks.pipeline.tasks.GenerateInternalLightningChunkTask;
+import org.terasology.world.chunks.pipeline.tasks.NotifyChunkTask;
 import org.terasology.world.internal.ChunkViewCore;
 import org.terasology.world.internal.ChunkViewCoreImpl;
-import org.terasology.world.propagation.light.InternalLightProcessor;
-import org.terasology.world.propagation.light.LightMerger;
 
 import java.math.RoundingMode;
 import java.util.Collection;
-import java.util.Collections;
 import java.util.Comparator;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.BlockingQueue;
 
 /**
- *
+ * Provides chunks received from remote source.
+ * <p>
+ * Loading/Unload chunks dependent on {@link org.terasology.network.Server}
+ * <p/>
+ * Produce events:
+ * <p>
+ * {@link OnChunkLoaded} when chunk received from server and processed.
+ * <p>
+ * {@link BeforeChunkUnload} when {@link org.terasology.network.Server} send invalidate chunk and chunk removing
  */
-public class RemoteChunkProvider implements ChunkProvider, GeneratingChunkProvider {
+public class RemoteChunkProvider implements ChunkProvider {
 
     private static final Logger logger = LoggerFactory.getLogger(RemoteChunkProvider.class);
-    private Map<Vector3i, Chunk> chunkCache = Maps.newHashMap();
-    private final BlockingQueue<Chunk> readyChunks = Queues.newLinkedBlockingQueue();
     private final BlockingQueue<Vector3i> invalidateChunks = Queues.newLinkedBlockingQueue();
-    private List<Chunk> sortedReadyChunks = Lists.newArrayList();
+    private Map<Vector3i, Chunk> chunkCache = Maps.newHashMap();
     private ChunkReadyListener listener;
     private EntityRef worldEntity = EntityRef.NULL;
 
     private BlockManager blockManager;
 
-    private ChunkGenerationPipeline pipeline;
-
-    private LightMerger<Chunk> lightMerger = new LightMerger<>(this);
+    private ChunkProcessingPipeline loadingPipeline;
 
     private LocalPlayer localPlayer;
 
     public RemoteChunkProvider(BlockManager blockManager, LocalPlayer localPlayer) {
         this.blockManager = blockManager;
         this.localPlayer = localPlayer;
-        pipeline = new ChunkGenerationPipeline(new ChunkTaskRelevanceComparator());
+        loadingPipeline = new ChunkProcessingPipeline(new ChunkTaskRelevanceComparator());
+
+        loadingPipeline.addStage(GenerateInternalLightningChunkTask::new)
+                .addStage(DeflateChunkTask::new)
+                .addStage(new LightMergerChunkTaskProvider(this, loadingPipeline))
+                .addStage(NotifyChunkTask.stage("Notify listeners", (chunk) -> {
+                    listener.onChunkReady(chunk.getPosition());
+                    worldEntity.send(new OnChunkLoaded(chunk.getPosition()));
+                    Chunk oldChunk = chunkCache.put(chunk.getPosition(), chunk);
+                    if (oldChunk != null) {
+                        oldChunk.dispose();
+                    }
+                    chunk.markReady();
```

Comment:
>this NotifyChunkTask has stuff in it not related to notifications: chunkCache and markReady.

... :(

>it calls something named onChunkReady before doing chunk.markReady? 

yeah it is some kind of smell. listener(Server) not need chunk itself now.

>chunk.markReady seems extraneous? i.e. is this task doing anything to improve the readiness of the chunk?

It is only `this.ready = true` T_T idk why do there :D

> could listener be null? IntelliJ surprisingly hasn't yelled at me about this, but it's not immediately obvious that subscribe will always be called with a non-null value before the pipeline receives a chunk.

Currently - no

>the Pipeline already manages its list of chunkTaskListeners, but here we have this NotifyChunkTask that is a different way of supporting listeners?

`ChunkTaskListeners` it is more internal listeners or technical. it firing on every stage.
`NotifyChunkTask` - Better naming `LambdaChunkTask` :D

>loadingPipeline.addListener((chunkTask) -> listener.onChunkReady(chunkTask.getPosition()))  // also needs null-check for `listener`
    .addListener((chunkTask) -> worldEntity.send(new OnChunkLoaded(chunkTask.getPosition()))

`listeners` - technical listeners. ChunkRemoveFromPipelineListener(aka ChunkTaskListener) It can be used for watch pipeline stage in other world. it is used by Stages itself.